app/services/base_services/base_contracted.rb

```
@@ -69,19 +69,19 @@ def perform(params = {})
     end
 
     def extract(params, attribute)
-      params = params ? params.dup : {}
-      [params, params.delete(attribute)]
+      params ||= {}
+      [params, params[attribute]]
```

🔴 here is now a functional difference. the key `attribute` was previously no longer part of the first element in the returned array. After your change, now it is still part of it.

in this example it means, that `send_notifications` is still part of `params`.

But I feel the need to change the old code which kicks concepts of immutability of method parameters. ^^ Maybe this would be a better line?

```ruby
[params.except(attribute), params[attribute]]
```

---

Edit: if this change WAS intentional, I couldn't find the reason in the previous comments, but please, ignore this comment then.

---

I remember changing this, the change was intentional. It was one of those cases where it was impossible to not change the behaviour.

Previously the `params` could've been accessed in two ways, via the attribute accessor or via the argument passed downwards. Now this method did behave differently for both of them: For the attribute accessor the key was left in the hash unchanged (due to `params.dup`), while for accessors of the argument it was left out (due to the return value of this method being passed down).

Now that we reduced it to one way of accessing params, I had to make an incompatible way to one of the ways accessing params. I guess this change is the less risky one:

* Code not expecting the parameter (but now unexpectedly having it) is either going to ignore its presence or to explode visibly, but there is a low chance of silent errors due to this
* contrary though, code expecting a parameter that we now never pass, might just assume that the param wasn't passed in the first place, failing silently
* In general (:saluting_face:) though, many services never get that param passed at all and will care neither way

---

Sounds fine for me, thanks for the explanation.